### PR TITLE
Add MSVC support via a distributed logger

### PR DESCRIPTION
**Pull request by erikmav**
Support MSVC by turning on link.exe unused-library flags and parsing output using a distributed logger that splits the libraries by default import libs and other libs.

### CODE REVIEW

**Comment by erikmav** [comment]
For #26

File: README.md [hunk 1]
```diff
@@ -2,18 +2,83 @@
...
+    <!-- Suppress XML doc comment issues to avoid errors during ReferenceTrimmer:
+         - CS0419: Ambiguous reference in cref attribute
+         - CS1570: XML comment has badly formed XML
+         - CS1573: Parameter has no matching param tag in the XML comment
+         - CS1574: XML comment has cref attribute that could not be resolved
+         - CS1584: XML comment has syntactically incorrect cref attribute
+         - CS1591: Missing XML comment for publicly visible type or member
+         - SA1602: Enumeration items should be documented
+    -->
+    <WarningsNotAsErrors>$(WarningsNotAsErrors);419;1570;1573;1574;1584;1591;SA1602;RT0002;RT0003</WarningsNotAsErrors>
```
Comment (dfederm) [inline README.md:32]: Remove `RT0002;RT0003`?

Comment (dfederm) [inline README.md:32]: `NoWarn` should be sifficient

File: src/Package/ReferenceTrimmer.Package.csproj [hunk 2]
```diff
@@ -34,6 +35,7 @@
       <Pack>true</Pack>
       <PackagePath>tools</PackagePath>
     </Content>
+    <None Include="..\..\README.md" Pack="true" PackagePath="\" />
```
Comment (dfederm) [inline src/Package/ReferenceTrimmer.Package.csproj:38]: Inconsistency with the content items above.

**Comment by dfederm** [comment]
Looks like the build is failing. Just tried against main and it seems to have the same problem. I'll take a look....

**Comment by dfederm** [comment]
Tests fixed in #65

**Comment by erikmav** [comment]
Marking this as Draft for now as @stan-sz mentioned a custom logger approach that might be a better approach to capturing stdout and taking action.

**Comment by dfederm** [comment]
In that case I'll steal the fix for #53 ;)

**Comment by erikmav** [comment]
By all means. In fact we could commit this which would spam vcxproj users with new output, then follow with a custom logger approach to then indicate what to fix (similar to the .md update in this PR) and which projects to remove to reduce complexity.

File: src/Loggers/InternalsVisibleTo.cs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("ReferenceTrimmer.E2ETests")]
```
Comment (dfederm) [inline src/Loggers/InternalsVisibleTo.cs:3]: This should be done in the project file

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -28,6 +28,7 @@ jobs:
         run: dotnet build --configuration Release --no-restore -bl:logs/build.binlog
 
       - name: Test
+        uses: ilammy/msvc-dev-cmd@v1  # Enable MSVC on Windows for test suite.
```
Comment (dfederm) [inline .github/workflows/ci.yml:31]: It's complaining that you can't have both a run and uses field on the same item

Comment (dfederm) [reply .github/workflows/ci.yml:31]: (that's why the PR gate is no longer running)

Comment (dfederm) [inline .github/workflows/ci.yml:31]: Pretty sure all this does is run `vcvarsall.bat`. It assumes VS is installed on the machine, which it won't be in GH Actions.

Comment (erikmav) [reply .github/workflows/ci.yml]: VS is available on the Windows agent. Prefer to keep this script in place as it no-ops on Linux automatically.

**Comment by erikmav** [comment]
@dfederm Ready for re-review

File: src/Loggers/MSVC/ForwardingLogger.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
+using Microsoft.Build.Framework;
+
+namespace ReferenceTrimmer.Loggers.MSVC;
+
+/// <summary>
+/// Forwarding logger that runs in each MSBuild worker node outside of the primary node.
+/// </summary>
+public sealed class ForwardingLogger : IForwardingLogger
+{
+    private readonly List<object> _loggerList = new();
```
Comment (dfederm) [inline src/Loggers/MSVC/ForwardingLogger.cs:10]: I'm not clear what this is for

File: src/Loggers/MSVC/UnusedLibsLogger.cs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        "odbc32.lib",
+        "odbccp32.lib",
+        "ole32.lib",
+        "oleaut32.lib",
+        "shell32.lib",
+        "user32.lib",
+        "uuid.lib",
+        "winspool.lib",
+    };
+
```
Comment (dfederm) [inline src/Loggers/MSVC/UnusedLibsLogger.cs:34]: Nit: extra newline

File: src/Loggers/MSVC/UnusedLibsLogger.cs [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    public void Dispose()
+    {
+        _eventSource.TaskStarted -= OnTaskStarted;
+        _eventSource.TaskFinished -= OnTaskFinished;
+        _eventSource.MessageRaised -= OnMessageRaised;
+    }
+
+    private void OnTaskStarted(object sender, TaskStartedEventArgs e)
+    {
+        if (!string.IsNullOrEmpty(e.ProjectFile) && e.TaskName == LinkTaskName)
```
Comment (dfederm) [inline src/Loggers/MSVC/UnusedLibsLogger.cs:81]: string comparisons should use `Equals`

File: src/Loggers/MSVC/ForwardingLogger.cs (hunk 5 above)
Comment (erikmav) [reply src/Loggers/MSVC/ForwardingLogger.cs:10]: The class itself or the _loggerList?

Comment (erikmav) [reply src/Loggers/MSVC/ForwardingLogger.cs]: Removed _loggerList in favor of a single ref to an UnusedLibsLogger.

Comment (dfederm) [reply src/Loggers/MSVC/ForwardingLogger.cs]: Still not sure I totally get why `UnusedLibsLogger` is separate. Perhaps just to separate the MSBuild stuff from the "real" logic?

Comment (dfederm) [reply src/Loggers/MSVC/ForwardingLogger.cs]: Doesn't make s huge different (one extra allocation at worst) so I guess it's whatever.

**Pull request merged**